### DISCUSSION TITLE
Best practice to reset another provider when its being watched by one that has been disposed?

### DISCUSSION THREAD

**Discussion by aidanmack** [rrousselGit/riverpod#3771] [Q&A]
So my original code looked a bit like this..
```@riverpod
class TestProvider extends _$TestProvider {

  Future<String> build() async {
    return getVehicleData();
  }

  Future<String> getVehicleData() async {
    return await ref.watch(apiServiceProvider).getVehicleData();
  }
}
```

The problem is that when TestProvider is autodisposed `apiService` is not refreshed/invalidated

So my fix is...

```@riverpod
class TestProvider extends _$TestProvider {
   late ApiService apiService;

  Future<String> build() async {
    apiService = ref.refresh(apiService)'
    return getVehicleData();
  }

  Future<String> getVehicleData() async {
    return await apiService.getVehicleData();
  }
}
```

Is that ok? I could not find a means to overide the autodispose and do the invalidate there.

**Comment by gdurandrexel** [comment]
If you really want a new state each time you watch the provider, you can make it a family provider, taking an Object as parameter, and pass a new Object each time you watch it. Be sure to make it autodispose though.
`ref.watch(apiServiceProvider(Object()))`
But that's an unusual use case.

Otherwise, you may want to ensure that apiProvider is also autodispose, in which case it will be disposed when TestProvider is disposed, and if no one else is watching it.
